packages/devtools_app/lib/src/memory/flutter/memory_controller.dart

```diff
@@ -507,31 +680,207 @@ class MPChartData {
     used.clear();
     capacity.clear();
     externalHeap.clear();
+    _pruning = false;
+  }
+}
+
+/// Prepare Engine (ADB memory info) data to plot in MPChart.
+class MPEngineChartData {
+  bool _pruning = false;
+
+  /// Datapoint entries for each Java heap value.
+  final List<Entry> javaHeap = <Entry>[];
+
+  /// Datapoint entries for each native heap value.
+  final List<Entry> nativeHeap = <Entry>[];
+
+  /// Datapoint entries for code size value.
+  final List<Entry> code = <Entry>[];
+
+  /// Datapoint entries for stack size value.
+  final List<Entry> stack = <Entry>[];
+
+  /// Datapoint entries for graphics size value.
+  final List<Entry> graphics = <Entry>[];
+
+  /// Datapoint entries for other size value.
+  final List<Entry> other = <Entry>[];
+
+  /// Datapoint entries for system size value.
+  final List<Entry> system = <Entry>[];
+
+  /// Datapoint entries for total size value.
+  final List<Entry> total = <Entry>[];
+
+  /// Prune entries plotted based on number of sample
+  /// minutes to display.
+  void prune(int minutesToDisplay) {
+    assert(minutesToDisplay > 0);
+
+    final len = javaHeap.length;
+
+    // Entries should match.
+    // All entries lengths are the same.
+    assert(len == javaHeap.length &&
+        len == nativeHeap.length &&
+        len == code.length &&
+        len == stack.length &&
+        len == graphics.length &&
+        len == other.length &&
+        len == system.length &&
+        len == total.length);
+    // All entries starting timestamps match,
+    assert(javaHeap.isNotEmpty &&
+        javaHeap[0].x == nativeHeap[0].x &&
+        nativeHeap[0].x == code[0].x &&
+        code[0].x == stack[0].x &&
+        stack[0].x == graphics[0].x &&
+        graphics[0].x == other[0].x &&
+        other[0].x == system[0].x &&
+        system[0].x == total[0].x);
+    // All entries ending timestamps match.
+    assert(javaHeap.isNotEmpty &&
+        javaHeap[len - 1].x == nativeHeap[len - 1].x &&
+        nativeHeap[len - 1].x == code[len - 1].x &&
+        code[len - 1].x == stack[len - 1].x &&
+        stack[len - 1].x == graphics[len - 1].x &&
+        graphics[len - 1].x == other[len - 1].x &&
+        other[len - 1].x == system[len - 1].x &&
+        system[len - 1].x == total[len - 1].x);
+
+    // Compute a new starting index from length - N minutes.
+    final timeLastSample = javaHeap.last.x;
+    var index = len - 1;
+    for (; index > 0; index--) {
```

Comment:
same comment here about for loop syntax. Also, this code is duplicated - please put into a helper

Reply:
This was re-written and obsolete code has been eliminated (forgot about a couple of methods that are never used).